### PR TITLE
Split Not, And, and Or into logical and bitwise versions

### PR DESCRIPTION
**Pull request by jacobhinkle**
Currently we have `UnaryOpType::Not` which implements both `!` and `~` depending on input types, and similar for `BinaryOpType::And` and `BinaryOpType::Or`, which implement `&&`/`&` and `||`/`|`. This can lead to unintuitive bugs when types are implicitly cast between `int` and `bool` (see #688). It is much clearer to split these operators, which this PR does. In summary, we remove the following operation types:
- `UnaryOpType::Not`
- `BinaryOpType::And`
- `BinaryOpType::Or`
- `BinaryOpType::Xor`

and we introduce these new op types:
- `UnaryOpType::LogicalNot`
- `UnaryOpType::BitwiseNot`
- `BinaryOpType::LogicalAnd`
- `BinaryOpType::BitwiseAnd`
- `BinaryOpType::LogicalOr`
- `BinaryOpType::BitwiseOr`
- `BinaryOpType::BitwiseXor`

We also introduce the following python ops:
- `fd.ops.logical_not`
- `fd.ops.logical_and`
- `fd.ops.logical_or`

Note that we do not implement `LogicalXor` which for boolean inputs is simply `x != y`.

Each of the bitwise operations, if receiving all `bool` inputs, will convert to the equivalent logical operation which will return a `bool`. If any argument is int then both arguments are cast to int and an int is returned.

Each of the logical operations, if receiving non-`bool` inputs, will first cast arguments to `bool`.

 This PR also bumps the version since we are making a minor change to the python interface. 

This PR replaces #688.

### CODE REVIEW

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
Torch and numpy both allow bitwise ops to work on bool values and their behavior works as if the bitwise ops were translated to logical ops. They also allow logical ops to work on int values, in which case it is as if the bitwise ops were first cast to bool. We could do this for consistency.

**Comment by jacobhinkle** [comment]
> Torch and numpy both allow bitwise ops to work on bool values and their behavior works as if the bitwise ops were translated to logical ops. They also allow logical ops to work on int values, in which case it is as if the bitwise ops were first cast to bool. We could do this for consistency.

I pushed a change that implements this. So the ops now behave similarly to before, but the IR represents the correct operation, which means we can properly generate C++ and evaluate expressions.

**Comment by jacobhinkle** [comment]
!build

File: csrc/codegen.cpp [hunk 1]
```diff
@@ -611,12 +611,7 @@ class CudaKernelGenerator : private kir::ConstIrVisitor {
     }
 
     if (auto op = inline_op_str(op_type)) {
-      if (alsoBooleanOperator(op_type) &&
-          uop->out()->dtype() == DataType::Bool) {
-        code_ << stringifyBooleanOp(op_type) << gen(uop->in());
-      } else {
-        code_ << *op << gen(uop->in());
-      }
+      code_ << *op << gen(uop->in());
```
Comment (zasdfgbnm) [inline csrc/codegen.cpp:614]: nice cleanup!

File: csrc/ir/builder.cpp [hunk 2]
```diff
@@ -83,19 +83,19 @@ Val* IrBuilder::negExpr(Val* val) {
...
   TORCH_CHECK(val != nullptr, "val is a nullptr in notExpr.");
   auto result = newScalar(val->dtype());
-  IrBuilder::create<UnaryOp>(UnaryOpType::Not, result, val);
+  IrBuilder::create<UnaryOp>(UnaryOpType::LogicalNot, result, val);
   return result;
 }
 
 Val* IrBuilder::derefExpr(Val* val) {
-  TORCH_CHECK(val != nullptr, "val is a nullptr in notExpr.");
+  TORCH_CHECK(val != nullptr, "val is a nullptr in derefExpr.");
```
Comment (zasdfgbnm) [inline csrc/ir/builder.cpp:91]: Thanks for cleaning up :-D

File: csrc/ir/nodes.cpp [hunk 3]
```diff
@@ -498,12 +502,7 @@ void BinaryOp::printHelper(
     ss << " " << inline_bop.value() << " ";
     ss << rhs;
   } else {
-    if (alsoBooleanOperator(op_type) &&
-        out()->getDataType().value() == DataType::Bool) {
-      ss << stringifyBooleanOp(op_type);
-    } else {
-      ss << op_type;
-    }
+    ss << op_type;
```
Comment (zasdfgbnm) [inline csrc/ir/nodes.cpp:505]: nice!

File: csrc/ops/arith.cpp [hunk 4]
```diff
@@ -455,28 +455,57 @@ NVFUSER_DEFINE_UNARY_OP(ceil, Ceil)
...
 NVFUSER_DEFINE_UNARY_OP(round, Round)
 NVFUSER_DEFINE_UNARY_OP(silu, Silu)
 NVFUSER_DEFINE_UNARY_OP(trunc, Trunc)
 NVFUSER_DEFINE_UNARY_OP(print, Print)
 #undef NVFUSER_DEFINE_UNARY_OP
 
+Val* logical_not(Val* v) {
+  if (!isBooleanType(v->dtype())) {
+    TORCH_CHECK(
+        isIntegralType(v->dtype()),
```
Comment (zasdfgbnm) [inline csrc/ops/arith.cpp:468]: I think this don't have to be integral, for example:
```C++
[cling]$ 1.0 && 2.0
input_line_3:2:9: warning: implicit conversion from 'double' to 'bool' changes value from 2 to true [-Wliteral-conversion]
 1.0 && 2.0
     ~~ ^~~
(bool) true
```

File: csrc/ops/arith.cpp [hunk 5]
```diff
@@ -455,28 +455,57 @@ NVFUSER_DEFINE_UNARY_OP(ceil, Ceil)
...
+        v->dtype());
+    v = castOp(DataType::Bool, v);
+  }
+  return unaryOp(UnaryOpType::LogicalNot, v);
+}
+
+TensorView* logical_not(TensorView* tv) {
+  if (!isBooleanType(tv->dtype())) {
+    TORCH_CHECK(
+        isIntegralType(tv->dtype()),
```
Comment (zasdfgbnm) [inline csrc/ops/arith.cpp:479]: Same here, don't have to be integral

**Review by zasdfgbnm — APPROVED** [review]
Thanks for the cleanup!

File: csrc/ops/arith.cpp (hunk 4 above)
Comment (jacobhinkle) [reply csrc/ops/arith.cpp:468]: Good point. I will remove those checks.

File: csrc/ir/builder.cpp [hunk 6]
```diff
@@ -366,7 +388,118 @@ Val* SimplifyingIrBuilder::andExpr(Val* lhs, Val* rhs) {
...
+  } else if (lhs_all_ones) {
+    return rhs_scalar;
+  } else if (rhs_all_ones) {
+    return lhs_scalar;
+  }
+
+  return IrBuilder::bitwiseAndExpr(lhs, rhs);
+}
+
+Val* SimplifyingIrBuilder::bitwiseOrExpr(Val* lhs, Val* rhs) {
```
Comment (jacobhinkle) [inline csrc/ir/builder.cpp:468]: Added these missing simplifying expressions.

**Comment by jacobhinkle** [comment]
With the latest changes, only new ops are added and the old ops don't break integration code. So I'm going to go ahead and merge.

**Pull request merged**